org.eclipse.dawnsci.analysis.dataset/src/org/eclipse/dawnsci/analysis/dataset/slicer/ROIVisitor.java

```diff
@@ -0,0 +1,80 @@
+/*-
+ * Copyright 2017 Diamond Light Source Ltd.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.dawnsci.analysis.dataset.slicer;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.dawnsci.analysis.api.roi.IROI;
+import org.eclipse.dawnsci.analysis.api.roi.IRectangularROI;
+import org.eclipse.january.dataset.IDataset;
+
+public class ROIVisitor {
+
+	public static void visitHorizontalSections(IDataset data, IROI inputROI, SliceVisitor visitor) throws Exception {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 34 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)